skimage/morphology/_flood_fill_cy.pyx

```diff
@@ -0,0 +1,143 @@
+#cython: cdivision=True
+#cython: boundscheck=False
+#cython: nonecheck=False
+#cython: wraparound=False
+
+"""Cython code used in _flood_fill.py."""
+
+cimport numpy as cnp
+
+
+# Must be defined to use QueueWithHistory
+ctypedef Py_ssize_t QueueItem
+
+include "../morphology/_queue_with_history.pxi"
+
+
+ctypedef fused dtype_t:
+    cnp.uint8_t
+    cnp.uint16_t
+    cnp.uint32_t
+    cnp.uint64_t
+    cnp.int8_t
+    cnp.int16_t
+    cnp.int32_t
+    cnp.int64_t
+    cnp.float32_t
+    cnp.float64_t
+
+
+# Definition of flag values used for `flags` in _flood_fill & _fill_plateau
+cdef:
+    # Border value - do not cross!
+    unsigned char BORDER = 2
+    # Part of the flood fill
+    unsigned char FILL = 1
+    # Not checked yet
+    unsigned char UNKNOWN = 0
+
+
+cpdef inline void _flood_fill_equal(dtype_t[::1] image,
```

Comment:
Is there a reason for using `cpdef` here? As long as this function isn't called somewhere in C- or Cython-code `cpdef` isn't needed and `inline void`, `nogil` are actually ineffective. So for readabilities sake simplifying the signature might give a clearer intention to the reader.

Reply:
@lagru 

Interesting! Do you have a reference as to nogil not working in cpdef'd functions?

Reply:
@jni As I understand it [`cpdef` will create two functions, equivalent to `def` and `cdef`](https://notes-on-cython.readthedocs.io/en/latest/function_declarations.html#cpdef-it-s-both) (note: I'm not sure whether that is an official document). So the second pure C-function is never used which won't do any harm apart from increasing the binary ever so slightly.

Concerning `nogil`: what I mean is that I'm not sure whether the `nogil` statement actually works together with the `def` part of the declaration (I have never seen it used that way). To add to this it clashes with the `with nogil` statement in the function body. So one of those isn't needed. For some reason I can't find the documentation to that keyword right now but I would recommend to just keep it as simple as possible. And as long as we don't need this functionality and keywords why risk it?

Reply:
Yes, you can use `with nogil` in a Cython function defined with `def`.  Here, the double `nogil` is because `cpdef` generates both the Python wrapper and the pure C-function - the C-function uses the declaration `nogil`, while the Python wrapper makes use of the `with nogil`. 

I generally use `cpdef` when reasonable, even if `def` would suffice at the time, to expose future possibilities.  We don't use it here right now, in this PR, but for the price of a function declaration we ship the ability for someone to be much faster if they want to call this from pure C.  And it's fast enough that it might be useful for real-time work, so I'm trying to be as sensitive to that as possible.

Reply:
> We don't use it here right now, in this PR, but for the price of a function declaration we ship the ability for someone to be much faster if they want to call this from pure C.

I would like this argument more if the function in question was part of the user-facing API. Right now I would raise an eyebrow if a user started to rely on this private function. 

I guess it comes down to different philosophies in this case.  While I'd still prefer a simpler approach until the we actually need to call this from Cython code, I can see no real disadvantages to yours. So to be clear, I'm fine to leave it as is. :)

Reply:
I have to side with @lagru here — although I very much like the idea of providing C-accessible functionality, if we go down that road we should do it with a public facing API, rather than semi-encourage users to use a private function that we might want to mess with in the future. In the meantime, we should aim to minimise our API surface.

I don't think this should hold up this PR, but wanted to use this chance to (maybe) establish a policy on this sort of thing.

Reply:
To be clear, I intend to use this function in this way in planned future work - relevant to interactive GUI as well as my research.  I'd very much like the pure C function to exist.

Reply:
Alright I'll allow it. But only because I like you. =P

Speaking of which, I'm not sure you ever gave me your guacamole recipe. =P And in looking for it in my email I just found a neglected to-do item — deprecating our deprecation policy! =P